File: 프로그래머스/3/42898. 등굣길/등굣길.py
def solution(m, n, puddles):
    answer = 0
    
    grid = [[0] * m for _ in range(n)]

    for pud in puddles:
        x, y = pud[1]-1, pud[0]-1   # x,y축 기준이 반대고 (1,1) 부터 시작하니깐 1씩 빼준다
        grid[x][y] = -1
    
    grid[0][0] = 1
    
    for i in range(n):
        for j in range(m):
            if i == 0 and j == 0: 
                continue
            
            # 현재 좌표가 물에 잠긴 경우 -> pass
            if grid[i][j] < 0:
                continue
                
            left, up = 0, 0 # 현재 좌표의 최단거리는 (좌측 좌표 최단거리) + (위측 좌표 최단거리) 임
            if j-1 < 0 or grid[i][j-1] < 0: # 왼쪽 좌표가 범위를 벗어나거나 물에 잠긴 경우 
                left = 0
            else:  
                left = grid[i][j-1]
                
            if i-1 < 0 or grid[i-1][j] < 0: # 위쪽 좌표가 범위를 벗어나거나 물에 잠긴 경우
                up = 0
            else:
                up = grid[i-1][j]
            
            
            grid[i][j] = left + up
            
             
    return grid[n-1][m-1] % 1000000007

File: 프로그래머스/3/42898. 등굣길/등굣길.java
import java.util.*;

class Solution {
    
    static int X = Integer.MAX_VALUE;
    static class Point {
        int x;
        int y;
        
        Point(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }
    
    public int solution(int m, int n, int[][] puddles) {
        
        int[][] dp = new int[n][m];
        for (int i = 0; i < puddles.length; i++) {
            dp[puddles[i][1]-1][puddles[i][0]-1] = X;
        }
        
        // init
        for (int j = 1; j < m; j++) {
            if (dp[0][j] == X) break;
                dp[0][j] = 1;
        }
        for (int i = 1; i < n; i++) {
            if (dp[i][0] == X) break;
                dp[i][0] = 1;
        }
        
        for (int i = 1; i < n; i++) {
            for (int j = 1; j < m; j++) {
                if (dp[i][j] == X) continue;
                
                int left = dp[i][j-1] != X ? dp[i][j-1] : 0;
                int up = dp[i-1][j] != X ? dp[i-1][j] : 0;
                dp[i][j] = (left + up) % 1000000007;
            }
        }
                
        return dp[n-1][m-1];
    }
}